## Document Loading in Langchain

### Text Loader

In [2]:
# Import libraries
from langchain.document_loaders import TextLoader

loader = TextLoader("nvdia news.txt")
data = loader.load()
data

[Document(metadata={'source': 'nvdia news.txt'}, page_content='Nvidia shares plunge amid $5.5bn hit over export rules to China\n\nShares in Nvidia plunged on Wednesday after the computer chip giant said it would be hit by $5.5bn (£4.2bn) in costs due to the US government tightening its export rules to China.\n\nThe firm, which has been at the heart of the artificial intelligence (AI) boom, will require licences to export its H20 AI chip to China, which has been one of its most popular.\n\nThe rules come as part of an escalating trade war between the US and China, with both countries introducing steep trade tariffs on each other covering various goods.\n\nNvidia shares fell almost 7% on Wednesday. The Nasdaq exchange it is listed on ended the day down 3.1%.\n\nThe company announced on Tuesday that the US government had told it last week that the H20 chip required a permit to be sold to China, including Hong Kong.\n\nThe tech giant said federal officials had advised them the licence req

In [3]:
type(loader)

langchain_community.document_loaders.text.TextLoader

In [6]:
loader.file_path

'nvdia news.txt'

In [7]:
data[0].metadata

{'source': 'nvdia news.txt'}

### CSV Loader

In [8]:
# Import libraries
from langchain.document_loaders import CSVLoader
import nltk
import pandas as pd

loader_CSV = CSVLoader("movies.csv")
data_CSV = loader_CSV.load()
data_CSV

[Document(metadata={'source': 'movies.csv', 'row': 0}, page_content='movie_id: 101\ntitle: K.G.F: Chapter 2\nindustry: Bollywood\nrelease_year: 2022\nimdb_rating: 8.4\nstudio: Hombale Films\nlanguage_id: 3\nbudget: 1\nrevenue: 12.5\nunit: Billions\ncurrency: INR'),
 Document(metadata={'source': 'movies.csv', 'row': 1}, page_content='movie_id: 102\ntitle: Doctor Strange in the Multiverse of Madness\nindustry: Hollywood\nrelease_year: 2022\nimdb_rating: 7\nstudio: Marvel Studios\nlanguage_id: 5\nbudget: 200\nrevenue: 954.8\nunit: Millions\ncurrency: USD'),
 Document(metadata={'source': 'movies.csv', 'row': 2}, page_content='movie_id: 103\ntitle: Thor: The Dark World\nindustry: Hollywood\nrelease_year: 2013\nimdb_rating: 6.8\nstudio: Marvel Studios\nlanguage_id: 5\nbudget: 165\nrevenue: 644.8\nunit: Millions\ncurrency: USD'),
 Document(metadata={'source': 'movies.csv', 'row': 3}, page_content='movie_id: 104\ntitle: Thor: Ragnarok\nindustry: Hollywood\nrelease_year: 2017\nimdb_rating: 7.9\

In [9]:
table = pd.read_csv("movies.csv")
table.head(2)

movie_id                                        title   industry  \
0       101                             K.G.F: Chapter 2  Bollywood   
1       102  Doctor Strange in the Multiverse of Madness  Hollywood   

   release_year  imdb_rating          studio  language_id budget revenue  \
0          2022          8.4   Hombale Films            3      1    12.5   
1          2022          7.0  Marvel Studios            5    200   954.8   

       unit currency  
0  Billions      INR  
1  Millions      USD

In [10]:
len(data_CSV)

9

In [11]:
data_CSV[0].metadata

{'source': 'movies.csv', 'row': 0}

In [12]:
loaderCSV2 = CSVLoader("movies.csv", source_column= "title")
dataCSV2 = loaderCSV2.load()
dataCSV2

[Document(metadata={'source': 'K.G.F: Chapter 2', 'row': 0}, page_content='movie_id: 101\ntitle: K.G.F: Chapter 2\nindustry: Bollywood\nrelease_year: 2022\nimdb_rating: 8.4\nstudio: Hombale Films\nlanguage_id: 3\nbudget: 1\nrevenue: 12.5\nunit: Billions\ncurrency: INR'),
 Document(metadata={'source': 'Doctor Strange in the Multiverse of Madness', 'row': 1}, page_content='movie_id: 102\ntitle: Doctor Strange in the Multiverse of Madness\nindustry: Hollywood\nrelease_year: 2022\nimdb_rating: 7\nstudio: Marvel Studios\nlanguage_id: 5\nbudget: 200\nrevenue: 954.8\nunit: Millions\ncurrency: USD'),
 Document(metadata={'source': 'Thor: The Dark World', 'row': 2}, page_content='movie_id: 103\ntitle: Thor: The Dark World\nindustry: Hollywood\nrelease_year: 2013\nimdb_rating: 6.8\nstudio: Marvel Studios\nlanguage_id: 5\nbudget: 165\nrevenue: 644.8\nunit: Millions\ncurrency: USD'),
 Document(metadata={'source': 'Thor: Ragnarok', 'row': 3}, page_content='movie_id: 104\ntitle: Thor: Ragnarok\nindus

In [13]:
dataCSV2[0].metadata

{'source': 'K.G.F: Chapter 2', 'row': 0}

### Unstructured URL Loader

In [14]:
# Import libraries
from langchain.document_loaders import UnstructuredURLLoader

loaderURL = UnstructuredURLLoader(urls=[
    "https://www.business-standard.com/markets/news/hdfc-bank-may-pocket-in-rs-9-373-cr-profit-from-hdb-financial-services-ipo-125062000535_1.html",
    "https://economictimes.indiatimes.com/industry/transportation/airlines-/-aviation/air-india-bookings-down-20-fares-dip-up-to-15-after-dreamliner-crash/articleshow/121977625.cms?from=mdr"
])

In [15]:
dataURL = loaderURL.load()
len(dataURL)

2

In [16]:
dataURL[0]

Document(metadata={'source': 'https://www.business-standard.com/markets/news/hdfc-bank-may-pocket-in-rs-9-373-cr-profit-from-hdb-financial-services-ipo-125062000535_1.html'}, page_content='Access Denied\n\nYou don\'t have permission to access "http://www.business-standard.com/markets/news/hdfc-bank-may-pocket-in-rs-9-373-cr-profit-from-hdb-financial-services-ipo-125062000535_1.html" on this server.\n\nReference #18.266d3f17.1750705249.19fb0255\n\nhttps://errors.edgesuite.net/18.266d3f17.1750705249.19fb0255')

## Text Splitting in Langchain

In [17]:
# Text from articles

text = """During lst years presidential campaign, Donald Trump constantly repeated his intention to bring about dramatic change as soon as he returned to the White House.
But few expected it to come at such breakneck speed.
In the three months since he took the oath of office, the 47th president has deployed his power in a way that compares to few predecessors.
In stacks of bound documents signed off with a presidential pen and policy announcements made in all caps on social media, his blizzard of executive actions has reached into every corner of American life.
To his supporters, the shock-and-awe approach has been a tangible demonstration of an all-action president, delivering on his promises and enacting long-awaited reforms.
But his critics fear he is doing irreparable harm to the country and overstepping his powers - crippling important government functions and perhaps permanently reshaping the presidency in the process.
Here are six turning points from the first 100 days.
A social media post sets off a constitutional firestorm
For once, it was not a Trump social media post that sparked an outcry.
Three weeks into the new term, at 10.13am on a Sunday morning, Vice-President JD Vance wrote nine words that signalled a strategy which has since shaped the Trump administrations second term.
'Judges arent allowed to control the executives legitimate power,' he declared on X.
In the media frenzy that followed, legal experts lined up to challenge that assertion, pointing to a 220-year-old principle which lies at the heart of American democracy.
Courts have the power to check and strike down any government action - laws, regulations and executive orders - they think violates the US Constitution.
Vances words represented a brazen challenge to judicial authority and, more broadly, the system of three co-equal branches of government crafted by Americas founders.
But Trump and his team remain unapologetic in extending the reach of the executive branch into the two other domains - Congress and the courts.
The White House has moved aggressively to wrest control of spending from Congress, unilaterally defunding programmes and entire agencies.
This erosion of its power has been largely met by silence on Capitol Hill, where Trumps Republicans hold slim majorities in both chambers.
The courts have been more resistant, with well over 100 rulings so far halting presidential actions they deem to be unconstitutional, according to a tally by the New York Times.
Some of the biggest clashes have been over Trumps immigration crackdown. In March, more than 200 Venezuelans deemed a danger to the US, were deported to El Salvador, many under sweeping wartime powers and without the usual process of evidence being presented in court.
A Republican-appointed judge on a federal appeals court said he was shocked by how the White House had acted.
'Now the branches come too close to grinding irrevocably against one another in a conflict that promises to diminish both,' Judge J. Harvie Wilkinson wrote.
Trump and White House officials have said they will obey court rulings, even as the president lambasts many of the judges who issue them and the administration at times moves slowly to fully comply.
It all amounts to a unique test of a constitutional system that for centuries has operated under a certain amount of good faith.
While Trump has been at the centre of this push, one of his principal agents of chaos is a man who wasnt born in the US, but who built a business empire there.
Day 32 - Government cuts
Brandishing a chainsaw, dressed in black
Elon Musk, dressed in black from head to toe and wearing sunglasses, stood centre stage and basked in the adulation of the Conservative Political Action Conference crowd.
The richest man in the world, who wants to cut trillions of dollars from the federal government, said he had a special surprise.
Argentinian President Javier Milei, known for his own budget-slashing, emerged from backstage and handed him a shiny gold chainsaw.
This is the chainsaw for bureaucracy, Musk exclaimed. CHAINSAW!!
It was a dramatic illustration not only of Musks enthusiasm for his "Department of Government Efficiency" (Doge) assignment, but also of the near rock-star status that the South African-born technologist has developed among the Trump faithful.
Since that appearance, Musk has dispatched his operatives across the federal government, pushing to access sensitive government databases and identify programmes to slash.
Although he has not come anywhere near to finding the trillions of dollars of waste he once promised, his cuts have drastically reduced dozens of agencies and departments - essentially shutting down the US Agency for International Development (USAID) and attempting to dismantle the Department of Education.
While pledges to cut "waste, fraud and abuse" in government and trim the ballooning federal deficit typically have broad appeal, the manner in which Musk has used his metaphorical chainsaw has led to conflict with senior government officials and stoked anger among some of the American public.
Some Trump supporters may approve of the administrations aggressive budget-cutting but other constituents have berated Republican legislators at town hall events.
Hecklers have expressed fear that the cuts will adversely affect popular government programmes like Social Security retirement plans, veterans benefits, and health insurance coverage for the poor and elderly.
Their concerns may not be entirely misplaced, given that these schemes make up the bulk of federal spending.
If these programmes are not cut back, sweeping tax cuts that Trump has promised would further increase the scale of US government debt and put at risk arguably his biggest election promise - economic prosperity.
Day 72 - Economy
'I had to think fast as billions was lost before my eyes'
When trader Richard McDonald saw Trump hold up his charts in the White House Rose Garden showing a list of countries targeted by US tariffs, he knew he had to act fast.
"I jumped to my feet because I wasnt expecting a board [of charts] - I was expecting an announcement," he says.
McDonald expected tariff cuts of 10% or 20%, but says "nobody expected these huge numbers".
He raced to understand which companies might be worst hit. Then he sold.
"There are billions being wiped off share prices every second, so it's really 'fastest finger first'."
He is one of the many traders who were at the coal face of global markets when share prices plunged everywhere following Trumps so-called "Liberation Day" tariff announcement.
The S&P 500 index of the largest firms listed in the US was hit particularly hard - and even though the White House has reversed course on some of the highest tariffs, it hasnt fully recovered since.
The economy was the biggest concern for US voters in Novembers election, and Trump rode a tide of deep unhappiness over Bidens handling of inflation all the way to victory.
His pledge to cut prices, pare back government regulation and boost homegrown industry was a pro-business message warmly welcomed on Wall Street and by many working Americans.
But as Trump tries to follow through on his promise of new tariffs, the economic costs, at least in the short term, have become painfully apparent.
The stock market is sinking, interest rates - including for home mortgages - are rising, and consumer confidence is down. Unemployment is also ticking up, in part due to the growing number of federal employees forced out of their jobs.
The Federal Reserve Bank, along with economic experts, warn Trumps plan will shrink economic growth and possibly lead to a recession.
While the presidents approval ratings on his handling of the economy have tumbled, many of his supporters are sticking with him. And in former industrial areas hollowed out by the loss of manufacturing jobs, there are hopes that tariffs could even the global playing field.
"Trump has earned back the respect," says truck driver Ben Maurer in Pennsylvania, referring to tariffs on China. "We are still the force to be reckoned with."
Economic concerns have contributed to Trumps overall decline in the polls, but in one key area, he is still largely on solid ground in the publics eye - immigration.
Immigration
"It's him! It's him! I recognise his features," says Myrelis Casique Lopez, pointing at a photo of men shackled and cuffed on the floor of one of the most infamous prisons in the world.
She had spotted her son in the image, taken from above, of a sea of shaven heads belonging to men in white T-shirts sat in long, straight rows.
At home in Maracay, Venezuela, Ms Casique was shown the photograph, first shared online by the El Salvador authorities, by a BBC reporter.
When she last had contact with her son, he was in the US and facing deportation to Venezuela but now he was 1,430 miles (2,300 km) away from her, one of 238 men sent by US authorities to a notorious mega-jail in El Salvador
The Trump administration says they are members of the Tren de Aragua gang - a powerful, multi-national crime operation - but Ms Casique insists her son is innocent.
A tough stance on immigration was a central plank of Trumps re-election campaign, and the president has used his broad powers of enforcement to deliver that pledge.
Illegal border crossings were falling at the end of the Biden presidency, but are now at their lowest monthly total for more than four years.
A majority of the US public still backs the crackdown, but it has had a chilling effect on communities of foreign students who have found themselves caught up in the blitz.
Some, including permanent residents, have been detained and face deportation because of their role in pro-Palestinian campus protests. They have rejected accusations that they support Hamas.
Civil rights lawyers warn that some migrants are being deported without due process, sweeping up the innocent among the "killers and thugs" that Trump says are being targeted.
While so far there havent been the level of mass deportations that some hoped for and others feared, newly empowered immigration enforcement agents have taken action across the US in businesses, homes and churches.
They have been active in universities too, which have become a prominent target of President Trump in several other ways.
Day 91 - War on institutions
A clash with academic, media and corporate worlds
On 21 April, Harvard Universitys president, Alan Garber, decided to confront the White House head-on.
In a letter to the university community, he announced a lawsuit challenging the Trump administrations move to freeze billions of dollars in federal grants.
It was, he said, an illegal attempt to "impose unprecedented and improper control" over Harvards operations.
The White House said it had to take action because Harvard had not tackled antisemitism on campus - an issue that Garber said the university was taking steps to address.
But the Ivy League colleges move was the most prominent display of resistance against Trumps use of presidential power to target American higher education, a longstanding goal energised by pro-Palestinian protests that engulfed campuses in 2024.
The president and his officials have since impounded or threatened to withhold billions of dollars in federal spending to reshape elite institutions like Harvard, which the president and many of his supporters think push a liberal ideology on students and researchers.
Earlier in the month, Columbia University in New York City had agreed to a number of White House demands, including changes to its protest policies, campus security practices and Middle Eastern studies department.
A similar dynamic has played out in the corporate and media worlds.
Trump has used the withholding of federal contracts as a way to pressure law firms to recruit and represent more conservatives.
Some of the firms have responded by offering the Trump administration millions of dollars in free legal services, while two firms have filed suit challenging the constitutionality of the administration punishments.
A defamation lawsuit Trump brought against ABC News has led to the media company contributing $15m (£11m) to Trumps presidential foundation.
CBS is also in talks to settle a separate lawsuit over a Kamala Harris interview, as its parent company Paramount seeks federal approval for a merger with Skydance Media.
The Associated Press, by contrast, has resisted administration pressure to accept Trumps "Gulf of America" name change despite the White Houses efforts to block the news agency from coverage of the president.
On the campaign trail, Trump warned about the runaway power of the federal government. Now in office, he is wielding that power in a way no previous modern president has attempted.
Nowhere, however, have the impacts of his efforts been more visible than within the federal government agencies and departments that he now controls.
Day 9 - Diversity projects
A retreat on race and identity
The press conference at the White House began with a moments silence for the victims of an aircraft collision over the Potomac River.
Within seconds of the pause coming to an end, however, Trump was on the attack.
A diversity and inclusion initiative at the Federal Aviation Agency was partly to blame for the tragedy, the president claimed, because it hired people with severe intellectual disabilities as air traffic controllers. He did not provide any evidence.
It was a startling moment that was emblematic of the attack his presidency has launched against inclusivity programmes that have proliferated in recent years across the US government and corporate world.
Trump has directed the federal government to end its diversity and equity (DEI) programmes and investigate private companies and academic institutions thought to be engaged in "illegal DEI".
His directive has accelerated moves among leading global companies like Meta and Goldman to cut back or eliminate these programmes.
First introduced in the 1960s in the wake of civil rights victories, early forms of DEI were an attempt to expand opportunities for black Americans. They later expanded to take in women, LGBT rights and other racial groups.
Efforts were stepped up and embraced by much of corporate America in the wake of the 2020 Black Lives Matters protests following the death of George Floyd at the hands of Minneapolis Police officers.
But to its critics, DEI was putting politics and race above talent, creating division and was no longer needed in modern America.
While Trumps directive seems to have support from a narrow majority of voters, some of the unexpected consequences have raised eyebrows.
Arlington National Cemetery scrubbed from its website all mentions of the history of black and female service members. And the Enola Gay aircraft that dropped an atomic bomb on Japan was initially flagged for removal from Pentagon documents, apparently due to the word "gay".
Donald Trumps first 100 days have been an unprecedented display of unilateral power exercised by a modern American president.
His efforts to dismantle large swaths of the federal government will take years, if not decades, for subsequent presidents to restore - if they so desire.
In other ways, however, Trump efforts so far may end up being less permanent. Without the support of new laws passed by Congress, many of his sweeping reforms could be wiped away by a future president.
And so to what extent this whirlwind start leads to lasting change remains an open question.
Later this year, the narrow Republican majorities in Congress will attempt to provide the legislative backing for Trumps agenda, but their success is far from guaranteed.
And in next years mid-term congressional elections, those majorities could be replaced by hostile Democrats bent on investigating the administration and curtailing his authority.
Meanwhile, more court battles loom - and while the US Supreme Court has a conservative tilt, its decisions on a number of key cases could ultimately cut against Trumps efforts.
The first 100 days of Trumps second term have been a dramatic show of political force, but the next 1,361 will be the real test of whether he can carve an enduring legacy."""

### Character Text Splitting

In [18]:
# Import libraries
from langchain.text_splitter import CharacterTextSplitter

splitter = CharacterTextSplitter(
    separator ='\n',  
    chunk_size = 400,
    chunk_overlap =0
)

chunks = splitter.split_text(text)

In [19]:
len(chunks)

51

In [20]:
for i in chunks:
    if (len(i)>400):
          print ("true")

### Recursive Character Text Splitting

In [21]:
# Import libraries
from langchain.text_splitter import RecursiveCharacterTextSplitter

rsplit = RecursiveCharacterTextSplitter(
        separators = ["\n\n", "\n", " "],
        chunk_size= 400,
        chunk_overlap =0
)

chunks_rsplit = rsplit.split_text(text)
len(chunks_rsplit)

51

In [22]:
for i in chunks_rsplit:
    if (len(i)>400):
          print ("true")

In [23]:
for i in chunks_rsplit:
    print (len(i))

353
374
380
276
323
310
276
177
378
355
354
299
196
243
171
307
293
371
395
372
351
372
323
369
273
387
329
362
329
314
366
365
307
278
245
268
281
342
311
389
341
330
394
355
329
136
275
280
294
349
348


### Process of forming Chunks

In [24]:
chunk_R1 = text.split("\n\n")

for i in chunk_R1:
    print(len(i))

16245


In [25]:
len(chunk_R1)

1

<p> There were no double lines entered(\n\n) in the text, hence only 1 chunk was made of length 16245

In [26]:
chunk_R2 = chunk_R1[0].split("\n")
len(chunk_R2)

107

In [27]:
for i in chunk_R2:
    print(len(i))

160
52
139
204
169
200
52
55
70
191
84
170
152
166
143
137
138
177
268
109
156
198
128
159
24
40
170
128
131
64
243
171
307
293
162
208
108
211
16
57
168
111
91
72
102
175
199
172
175
147
235
133
273
159
165
11
49
185
143
138
223
164
164
141
172
190
175
214
121
28
49
101
155
108
169
245
268
213
67
127
214
140
170
208
180
149
26
30
133
79
250
203
190
131
223
199
129
136
275
125
154
201
92
170
178
176
171


The final chunk size is size_sub_chunk_1 + ... + size_sub_chunk_n + n-1